`features/embeddings` and remove parts from search (#697)
spiceai/docs: spiceaidocs/docs/features/embeddings/index.md

---
title: 'Embedding Datasets'
sidebar_label: 'Embedding Datasets'
description: 'Learn how to define, or augment existing datasets with embedding column(s).'
sidebar_position: 11
pagination_prev: null
pagination_next: null
---

> # Embedding Datasets
> 
> Learn how to define and augment datasets with embedding columns for advanced search capabilities.

Reviewer: ```suggestion
Learn how to define and augment datasets with embedding columns for advanced search capabilities.
```

> ## Overview
> 
> Spice supports three methods for working with embedding columns within datasets:
> 
> 1. [**Just-in-Time (JIT) Embeddings**](#jit-embeddings): Computes embeddings for the dataset, on-demand, during query execution.
> 2. [**Accelerated Embeddings**](#accelerated-embeddings): Precompute embeddings by accelerating the source dataset.
> 3. [**Passthrough Embeddings**](#passthrough-embeddings): Uses the existing embeddings from the underlying source datasets.

Reviewer: ```suggestion
## Overview

Spice provides three distinct methods for handling embedding columns in datasets:

1. **[Just-in-Time (JIT) Embeddings](#jit-embeddings)**: Dynamically computes embeddings, on-demand, during query execution, without precomputing data.
   
2. **[Accelerated Embeddings](#accelerated-embeddings)**: Precomputes embeddings by transforming and augmenting the source dataset for faster query and search performance.
   
3. **[Passthrough Embeddings](#passthrough-embeddings)**: Utilizes pre-existing embeddings directly from the underlying source datasets, bypassing any additional computation.
```

## Configuring Embedding Models

> Before configuring dataset embeddings, you must define the embedding models in your `spicepod.yaml`, for example:

Reviewer 2: ```suggestion
Before configuring dataset embeddings define the embedding models in the `spicepod.yaml`, for example:
```

```yaml
embeddings:
  - name: local_embedding_model
    from: huggingface:huggingface.co/sentence-transformers/all-MiniLM-L6-v2

  - from: openai
    name: remote_service
    params:
      openai_api_key: ${ secrets:SPICE_OPENAI_API_KEY }
```

See [Embedding components](/components/embeddings/) for more information on embedding models.

## Embedding Methods
### Just-in-Time (JIT) Embeddings {#jit-embeddings}

> JIT embeddings are computed during query execution. This is useful when you can't or don't want to pre-compute embeddings (e.g. if the dataset is large, infrequently queried, has heavy prefiltering). To add an embedding column, specify it within the dataset's column.

Reviewer 2: ```suggestion
JIT embeddings are computed during query execution. This is useful when pre-computing embeddings is infeasible (e.g. if the dataset is large, infrequently queried, has heavy prefiltering). To add an embedding column, specify it within the dataset's column.
```

3. **Embeddings Column Data Type:**
   - The embeddings column must have the following [Arrow data type](reference/datatypes.md) when loaded into Spice:
     1. `FixedSizeList[Float32 or Float64, N]`, where `N` is the dimension (size) of the embedding vector. `FixedSizeList` is used for efficient storage and processing of fixed-size vectors.
     2. If the column is [**chunked**](#chunking), use `List[FixedSizeList[Float32 or Float64, N]]`.

4. **Offset Column for Chunked Data:**
   - If the underlying column is chunked, there must be an additional offset column named `<column_name>_offsets` with the following Arrow data type:
     1. `List[FixedSizeList[Int32, 2]]`, where each element is a pair of integers `[start, end]` representing the start and end indices of the chunk in the underlying text column. This offset column maps each chunk in the embeddings back to the corresponding segment in the underlying text column.
     - _For instance, `[[0, 100], [101, 200]]` indicates two chunks covering indices 0–100 and 101–200, respectively._

> By following these guidelines, you can ensure that your dataset with pre-existing embeddings is fully compatible with embedding functionalities provided by Spice.

Reviewer 2: ```suggestion
Following these guidelines ensures that the dataset with pre-existing embeddings is fully compatible with embedding functionalities provided by Spice.
```

## Advanced Configuration
### Chunking

Spice also supports chunking of content before embedding, which is useful for large text columns such as those found in [Document Tables](/components/data-connectors/index.md#document-support). Chunking ensures that only the most relevant portions of text are returned during search queries. Chunking is configured as part of the embedding configuration.

```yaml
datasets:
  - from: github:github.com/spiceai/spiceai/issues
    name: spiceai.issues
    acceleration:
      enabled: true
    columns:
      - name: body
        embeddings:
          - from: local_embedding_model
            chunking:
              enabled: true
              target_chunk_size: 512
```

The `body` column will be divided into chunks of approximately 512 tokens, while maintaining structural and semantic integrity (e.g. not splitting sentences). See the [API reference](/reference/spicepod/datasets#columns-embeddings-chunking) for full details.

#### Row Identifiers

> Like a primary key, the `row_id` field specifies which column(s) uniquely identifies each row. This is useful for embedding datasets that don't have a primary key by default. This is important for chunked embedding datasets, so that operations (e.g. [`v1/search`](/api/http/search)), can be able to map multiple chunked vectors to a single dataset row. The `row_id` can be set in the `columns[*].embeddings[*].row_id`.
> ```yaml

Reviewer 2: ```suggestion
Like a primary key, the `row_id` field specifies which column(s) uniquely identifies each row. This is useful for embedding datasets that don't have a primary key by default. This is important for chunked embedding datasets, so that operations (e.g. [`v1/search`](/api/http/search)), can be able to map multiple chunked vectors to a single dataset row. The `row_id` can be set in the `columns[*].embeddings[*].row_id`.

```yaml
```
